src/main/java/com/codicesoftware/plugins/hudson/WorkspaceManager.java

```diff
@@ -56,15 +108,123 @@ public static void cleanWorkspace(PlasticTool tool, FilePath workspacePath, Clea
+
+        LOGGER.fine(String.format("Switching workspace at '%s' to '%s'", targetWorkspacePath, targetSpec));
+        SwitchCommand command = new SwitchCommand(targetWorkspacePath, targetSpec);
+        CommandRunner.execute(tool, command);
+    }
+
     public static String generateUniqueWorkspaceName() {
         return "jenkins_" + UUID.randomUUID().toString().replaceAll("-", "");
     }
+
+    private static void deleteOldWorkspacesIfNeeded(
+            PlasticTool tool,
+            FilePath workspacePath,
+            CleanupMethod cleanup,
+            List<Workspace> workspaces)
+            throws IOException, InterruptedException {
+
+        // Handle situation where workspace exists in child path.
+        List<Workspace> innerWorkspaces = findWorkspacesInsidePath(workspacePath, workspaces);
+        for (Workspace workspace : innerWorkspaces) {
+            deleteWorkspace(tool, workspace, workspaces);
+        }
+
+        // Handle situation where workspace exists in parent path.
+        List<Workspace> outerWorkspaces = findWorkspacesOutsidePath(workspacePath, workspaces);
+        for (Workspace workspace : outerWorkspaces) {
+            deleteWorkspace(tool, workspace, workspaces);
+        }
+
+        if (cleanup == CleanupMethod.DELETE) {
+            Workspace workspace = findWorkspaceByPath(workspacePath, workspaces);
+            if (workspace != null) {
+                deleteWorkspace(tool, workspace, workspaces);
+            }
+        }
+    }
+
+    protected static boolean isSameWorkspacePath(String actual, String expected) {
+        String actualFixed = actual.replaceAll("\\\\", "/");
+        String expectedFixed = expected.replaceAll("\\\\", "/");
+
+        Matcher windowsPathMatcher = WINDOWS_PATH_PATTERN.matcher(expectedFixed);
+        // Windows paths are case insensitive
+        if (windowsPathMatcher.matches()) {
+            return actualFixed.equalsIgnoreCase(expectedFixed);
+        }
+        return actualFixed.equals(expectedFixed);
+    }
+
+    private static void deleteWorkspace(
+            PlasticTool tool, Workspace workspace, List<Workspace> workspaces)
+            throws IOException, InterruptedException {
+        WorkspaceManager.deleteWorkspace(tool, workspace.getPath());
+        workspace.getPath().deleteContents();
+        workspaces.remove(workspace);
+    }
+
+    protected static Workspace findWorkspaceByPath(FilePath workspacePath, List<Workspace> workspaces) {
+        for (Workspace workspace : workspaces) {
+            if (isSameWorkspacePath(workspace.getPath().getRemote(), workspacePath.getRemote())) {
+                return workspace;
+            }
+        }
+        return null;
+    }
+
+    protected static List<Workspace> findWorkspacesInsidePath(FilePath workspacePath, List<Workspace> workspaces) {
+        return matchWorkspaces(workspaces, testPath -> isNestedWorkspacePath(workspacePath.getRemote(), testPath));
```

Comment:
Should we use a different name for this "testPath"?

Reply:
Fixed in https://github.com/jenkinsci/plasticscm-plugin/pull/63/commits/ae175e8b6d4ad5004711f72e70af2ec4af88dcc0